packages/ai-jsx/src/lib/llama2.tsx

```diff
@@ -0,0 +1,174 @@
+  /** The maximum number of tokens to generate */
+  max_length?: number;
+  /** Penalty for repeated words in the output. Must be in the range [0.01, 5]. */
+  repetition_penalty?: Llama2ModelProps['repetitionPenalty'];
+
+  top_p: ModelProps['topP'];
+}
+
+/**
+ * A fetcher function that points to your Llama2 endpoint.
+ *
+ * It should return a generator that yields chunks of the response.
+ *
+ * For example, if the model generates the string "hello world I am an AI", the generator might yield:
+ *  - hel
+ *  - lo world I a
+ *  - m an AI
+ *
+ * If you don't have a streaming API, you can just yield a single chunk.
+ */
+export type FetchLlama2 = (args: FetchLlama2Args) => AsyncGenerator<string, string>;
+
+export interface Llama2Props extends ModelPropsWithChildren, Pick<Llama2ModelProps, 'repetitionPenalty'> {
+  fetch: FetchLlama2;
+}
+
+/**
+ * If you use a Llama2 model without specifying the max tokens for the completion, this value will be used as the default.
+ */
+export const defaultMaxTokens = 500;
+
+/**
+ * This will polyfill a chat model call for Llama2. However, note that Llama2 is not a structured chat model
+ * in the same way GPT-4 is. You may be better off thinking about Llama2 as a traditional completion model.
+ * And as such, you may wish to use `<Completion>` instead of `<ChatCompletion>`, for finer-grained control.
+ *
+ * The key difference between Chat and Completion models is that Chat models are structured as a conversation
+ * between a user and an assistant, whereas Completion models are just trying to predict subsequent text. So
+ * if you give a Completion model a prompt like:
+ *
+ *    Human: Hi, how are you?
+ *    Assistant: I'm doing well, how are you?
+ *    Human: I'm good. What is your name?
+ *
+ * You may get a completion like:
+ *    Assistant: Sam. What's yours?
+ *    Human: Chris.
+ *
+ * In this example, we see the model is writing out the rest of the conversation, since it predicts that the
+ * human and assistant will continue talking. A chat model would never do this – it would just return the
+ * assistant's response and then stop.
+ *
+ * You may have to do a little more prompt engineering work to get the model in the mood to chat and stop
+ * responding after a single message.
+ */
+async function* Llama2ChatModel(props: Llama2Props, { render, logger }: AI.ComponentContext): AI.RenderableStream {
+  const messages = await getPolyfilledMessages(
+    props.children,
+    render,
+    'Llama2',
+    ErrorCode.Llama2DoesNotSupportFunctions
+  );
+  yield AI.AppendOnlyStream;
+  const prompt = messages.join('\n\n');
+  const llama2Args: FetchLlama2Args = {
+    prompt,
+    max_length: props.maxTokens ?? defaultMaxTokens,
+    repetition_penalty: props.repetitionPenalty,
+    temperature: props.temperature,
+    top_p: props.topP,
+  };
+  logger.debug({ llama2Args }, 'Calling Llama2');
+  const response = await props.fetch(llama2Args);
+  yield* response;
+  return AI.AppendOnlyStream;
```

Comment:
@petersalas is this the right thing to be doing?

Reply:
Yeah, this is consistent with the convention in the other places.  (Alternatively you could `return null` or `return ""`, but IMO `return AI.AppendOnlyStream` also indicates _why_ the return value is empty.)